File: Algebras/RegionInterpolation/Java/src/movingregion/SingleMatch.java


package movingregion;

public class SingleMatch
{
    private RegionTreeNode source;
    private RegionTreeNode[] targets;
    
    public SingleMatch(RegionTreeNode source,RegionTreeNode target)
    {
        this.source=source;
        targets=new RegionTreeNode[1];
        targets[0]=target;
    }
    
    public RegionTreeNode getSource()
    {
        return(source);
    }
    
    public int getNrTargets()
    {
        int res=0;
        for(int i=0;i<this.targets.length;i++)
        {
            if(targets[i]!=null)
            {
                res++;
            }
        }
        return(res);
    }
    
    public void removeNulls()
    {
        for(int i=0;i<this.targets.length;i++)
        {
            if(targets[i]==null)
            {
                this.removeTarget(i);
            }
        }
    }
    
    public RegionTreeNode getTargetAt(int i)
    {
        return(targets[i]);
    }
    
    public void removeTarget(int index)
    {
        RegionTreeNode[] newTargets=new RegionTreeNode[targets.length-1];
        int j=0;
        for(int i=0;i<targets.length;i++)
        {
            if(i!=index)
            {
                newTargets[j]=targets[i];
                j++;
            }
            
        }
        targets=newTargets;
    }
    
    public void removeTargets()
    {
        targets=targets=new RegionTreeNode[1];
        targets[0]=null;
    }
    
    public void addTarget(RegionTreeNode target)
    {
        RegionTreeNode[]tmp=new RegionTreeNode[targets.length+1];
        for(int i=0;i<targets.length;i++)
        {
            tmp[i]=targets[i];
        }
        tmp[targets.length]=target;
        targets=tmp;
    }
    
    public int hashCode()
    {
        return(source.hashCode());
    }
    
    public boolean equals(Object o)
    {
        if(o instanceof SingleMatch)
        {
            SingleMatch tmp=(SingleMatch)o;
            return(this.source.equals(tmp.getSource()));
        }
        else
        {
            if(o instanceof RegionTreeNode)
            {
                RegionTreeNode tmp=(RegionTreeNode)o;
                return(tmp.equals(this.source));
            }
            else
            {
                return false;
            }
        }
    }
    
    public String toString()
    {
        String res=this.source+"==>";
        if(targets==null)
        {
            res=res+"null";
        }
        else
        {
            for(int i=0;i<this.targets.length;i++)
            {
                res=res+'\n'+targets[i];
            }
        }
        return(res);
    }
}

File: Algebras/RegionInterpolation/SingleMatch.cpp
/*
 
 see SingleMatch.h for documentation
 
 \tableofcontents
 
*/

#include "RegionInterpolator.h"

namespace RegionInterpol
{
/*
 
1 Constructors and Destructor
 
*/    
SingleMatch :: SingleMatch(RegionTreeNode *src, RegionTreeNode *target)
{
   source = src;
   targets.push_back(target);
}
/*

1 Get functions

1.1 getSource()

*/   
RegionTreeNode *SingleMatch :: getSource()
{
    return(source);
}
/*

1.1 getNrTargets()

*/  
int SingleMatch :: getNrTargets()
{
    int res = 0;
    for(unsigned int i = 0; i < targets.size(); i++)
    {
        if(targets[i] != NULL)
        {
            res++;
        }
    }
    return(res);
}
/*

1.1 getTargetAt()

*/  
RegionTreeNode *SingleMatch :: getTargetAt(int i)
{
    return(targets[i]);
}
/*

1 Set functions

1.1 addTarget()

*/     
void SingleMatch :: addTarget(RegionTreeNode *target)
{
   bool found = false;
   for(unsigned int i = 0; i < targets.size(); i++)
   {     
      if(target->equals(targets[i]))
      {
         found = true;
         break;
      }
   }
   if(!found)
      targets.push_back(target);
}
/*

1.1 removeTarget()

*/     
void SingleMatch :: removeTarget(int index)
{
   targets.erase(targets.begin() + index);  
}
/*

1.1 removeTargets()

*/
void SingleMatch :: removeTargets()
{
   targets.clear();
}
/*

1.1 removeNulls()

*/
void SingleMatch :: removeNulls()
{
    for(unsigned int i = 0; i < targets.size(); i++)
    {
        if(targets[i] == NULL)
        {
            removeTarget(i);
        }
    }
}
/*

1 Public Methods

1.1 hashCode()

*/
int SingleMatch :: hashCode()
{
    return(source->hashCode());
}
/*

1.1 equals()

*/
bool SingleMatch :: equals(SingleMatch* o)
{            
   return(source->equals(o->getSource()));        
}
/*

1 Operators

1.1 $<<$

*/  
std::ostream& operator << (std::ostream& s, SingleMatch sm)
{
   s << "Match:" << std::endl;  
   s << sm.getSource() << "===============>" << std::endl;
   for(int i = 0; i < sm.getNrTargets(); i++)
   {
      s << sm.getTargetAt(i);
      if(i != sm.getNrTargets() - 1)
         s << "... ... ..." << std::endl;         
   }        
   s << std::endl;
   return s;
}  

}

